feat(discover): Add option to filter for project

wmak opened the pull request:
- This is to help improve usability when an organization have a lot of projects, if many of them have never sent a transaction we can just exclude those so the query we send to snuba is far smaller

wmak committed:
feat(discover): Add option to filter for project

- This is to help improve usability when an organization have a lot of
  projects, if many of them have never sent a transaction we can just
  exclude those so the query we send to snuba is far smaller

Zylphrex commented on src/sentry/search/events/builder/base.py:
```diff
@@ -570,7 +573,8 @@ def resolve_params(self) -> list[WhereType]:
         # Snuba will complain with UnqualifiedQueryError: validation failed for entity...
-        if not self.params.project_ids:
+        project_ids = self.resolve_projects()
+        if not project_ids:
```
Could this result in poor UX? For example, users go on to a page like discover before sending an event then they get this error saying they need to select an project when that problem is that they need to send an event first.

wmak replied:
🤔 I think I'll just add `You need to select at least one project with data`

wmak committed:
change to rate based option

Zylphrex commented on src/sentry/search/events/builder/discover.py:
```diff
@@ -86,6 +87,12 @@ def resolve_field(self, raw_field: str, alias: bool = False) -> Column:
 
+    def resolve_projects(self) -> list[int]:
+        if options.get("sentry.search.events.project.check_event"):
+            return [proj.id for proj in self.params.projects if proj.flags.has_transactions]
```
Can't discover also query for errors?

wmak replied:
Good callout, changed this to treat the condition differently based on the dataset selected
Discover -> either an error or a txn has been sent
Events -> an error has been sent
Txn/IssuePlatform -> a txn has been sent

wmak committed:
fix: Address PR comments

- Expand error message
- treat different datasets differently for resolve-projects
- fix a test that assumed that project has_transactions is false by
  default

wmak committed:
fix: move create_project to just snuba tests

- Remove changes to other tests

wmak committed:
fix test

wmak committed:
ref: Move the error to resolve_projects

- this is so we can give a better error message, and also fallback to
  resolve_project if somehow we're querying on a weird dataset
